File: drafting_service/core/edit_logger.py
"""Recruiter edit logging & similarity metric calculator."""

from __future__ import annotations

import difflib
from typing import Any, Dict

from logger import get_logger

log = get_logger(__name__)


def log_recruiter_edit(draft_id: str, original_body: str, edited_body: str) -> Dict[str, Any]:
    """Calculate similarity ratio and edit distance between AI draft and recruiter edited text."""
    orig = original_body or ""
    edited = edited_body or ""

    matcher = difflib.SequenceMatcher(None, orig, edited)
    similarity = matcher.ratio()  # 1.0 = identical, 0.0 = completely changed
    edit_pct = round((1.0 - similarity) * 100, 2)

    result = {
        "draft_id": draft_id,
        "original_length": len(orig),
        "edited_length": len(edited),
        "similarity_score": round(similarity, 4),
        "edit_percentage": edit_pct,
        "was_edited": edit_pct > 0.0,
    }

    log.info("Recruiter edit logged for %s: edit_pct=%.2f%% (similarity=%.4f)", draft_id, edit_pct, similarity)
    return result

File: server/src/drafting/editLogger.ts
/** Recruiter edit logging & similarity metric calculator.
 * Direct port of drafting_service/core/edit_logger.py. Exported as a plain
 * function, not wired to a route -- POST /draft/edit-log had zero live
 * callers when this was ported (confirmed via repo-wide grep); this stays
 * available for a future recruiter-edit-tracking feature to call directly. */

import { sequenceMatcherRatio } from "./lib/sequenceMatcher";

export interface EditLogResult {
  draft_id: string;
  original_length: number;
  edited_length: number;
  similarity_score: number;
  edit_percentage: number;
  was_edited: boolean;
}

export function logRecruiterEdit(draftId: string, originalBody: string, editedBody: string): EditLogResult {
  const orig = originalBody || "";
  const edited = editedBody || "";

  const similarity = sequenceMatcherRatio(orig, edited);
  const editPct = Math.round((1.0 - similarity) * 100 * 100) / 100;

  return {
    draft_id: draftId,
    original_length: orig.length,
    edited_length: edited.length,
    similarity_score: Math.round(similarity * 10000) / 10000,
    edit_percentage: editPct,
    was_edited: editPct > 0.0,
  };
}
